PHP 8.0 and PHP 8.1 support, BetterReflection 5.x upgrade

Ocramius opened the pull request:
This patch brings in PHP 8.0 and 8.1 support through the massive work done by @kukulich on
`roave/better-reflection` over the past few months.

The upgrade is mostly smooth, but we have some rough edges with:

 * [x] `TypeIsCovariant`, which needs `ReflectionType`s as well as `Reflector`s
 * [x] `TypeIsContravariant`, which needs `ReflectionType`s as well as `Reflector`s
 * [x] `ReturnTypeCovarianceChanged` uses hacky reflection to get a `Reflector` from a `ReflectionFunctionAbstract` (internal state!) - needs https://github.com/Roave/BetterReflection/pull/836
 * [x] `ParameterTypeContravarianceChanged` uses hacky reflection to get a `Reflector` from a `ReflectionFunctionAbstract` (internal state!) - needs https://github.com/Roave/BetterReflection/pull/836
 * [x] we miss tests for covariance/contravariance with `ReflectionUnionType` and `ReflectionIntersectionType` (`infection/infection` will highlight that)
 * [x] https://github.com/Roave/BetterReflection/pull/902
 * [x] tests for covariance/contravariance of nullable `ReflectionIntersectionType` (needs https://github.com/Roave/BetterReflection/pull/902 )

Fixes #283

Ocramius committed:
PHP 8 support and BetterReflection 5.x upgrade

This patch brings in PHP 8.0 and 8.1 support through the massive work done by @kukulich on
`roave/better-reflection` over the past few months.

The upgrade is mostly smooth, but we have some rough edges with:

 * [x] `TypeIsCovariant`, which needs `ReflectionType`s as well as `Reflector`s
 * [x] `TypeIsContravariant`, which needs `ReflectionType`s as well as `Reflector`s
 * [ ] `ReturnTypeCovarianceChanged` uses hacky reflection to get a `Reflector` from a `ReflectionFunctionAbstract` (internal state!)
 * [ ] `ParameterTypeContravarianceChanged` uses hacky reflection to get a `Reflector` from a `ReflectionFunctionAbstract` (internal state!)
 * [ ] we miss tests for covariance/contravariance with `ReflectionUnionType` and `ReflectionIntersectionType` (`infection/infection` will highlight that)

Ocramius commented on composer.json:
```diff
@@ -2,13 +2,13 @@
-        "ocramius/package-versions": "^2.1.0",
-        "roave/better-reflection": "^4.12.2",
+        "ocramius/package-versions": "^2.4.0",
+        "roave/better-reflection": "5.0.x-dev@DEV",
```
We need a stable release of `roave/better-reflection` before merging this

Ocramius replied:
Will defer this to a new issue

Ocramius commented on composer.json:
```diff
@@ -45,6 +45,9 @@
-        "sort-packages": true
+        "sort-packages": true,
+        "platform": {
+            "php": "8.0.99"
```
Unsure if needed: will prevent invalid dependency upgrades though

WyriHaximus replied:
That is exactly how I'm using it 👍

Ocramius commented on src/CompareApi.php:
```diff
@@ -4,20 +4,20 @@
-        ClassReflector $definedSymbols,
-        ClassReflector $pastSourcesWithDependencies,
-        ClassReflector $newSourcesWithDependencies
+        Reflector $definedSymbols,
```
Loads of inherited BC breaks because of this, but they are very much acceptable

Ocramius commented on src/DetectChanges/BCBreak/FunctionBased/ParameterTypeChanged.php:
```diff
@@ -79,7 +79,6 @@ private function typeToString(?ReflectionType $type): string
 
-        return ($type->allowsNull() ? '?' : '')
-            . $type->__toString();
+        return $type->__toString();
```
`ReflectionType#__toString()` now includes nullability as per `roave/better-reflection:5.0.0` - this is a BC break that led to some simplification and also code changes all around the place.

Ocramius commented on src/DetectChanges/BCBreak/FunctionBased/ParameterTypeContravarianceChanged.php:
```diff
@@ -70,7 +78,18 @@ private function typeToString(?ReflectionType $type): string
+    }
+
+
+    /** @TODO may the gods of BC compliance be merciful on me */
```
May the gods have mercy of my.

This is ugly, but it highlights a design limitation when trying to traverse a whole reflection graph.

Perhaps @kukulich has a better idea on how to achieve this, but the problem is basically that we're trying to find `parent` classes and related interfaces by their name, and it is not possible to do so from a `ReflectionType` from `roave/better-reflection:5.0.0`, as `ReflectionType` is mostly immutable.

Instead, we have to fetch the associated reflector here.

Injecting the reflector upfront is not feasible either here, with the current design (the reflector is built at runtime, by some services that do run a `composer install`) :thinking:

kukulich replied:
https://github.com/Roave/BetterReflection/pull/836

Ocramius commented on src/DetectChanges/BCBreak/FunctionBased/ReturnTypeCovarianceChanged.php:
```diff
@@ -55,7 +62,17 @@ private function typeToString(?ReflectionType $type): string
+        return $type->__toString();
+    }
+
+    /** @TODO may the gods of BC compliance be merciful on me */
```
Similar to above: this is currently the "cleanest" approach, but it highlights the limitation of the current `ReflectionType` design.

Ocramius commented on src/DetectChanges/Variance/TypeIsContravariant.php:
```diff
@@ -29,12 +35,17 @@ public function __invoke(
         }
 
+        if (! $type instanceof ReflectionNamedType || ! $comparedType instanceof ReflectionNamedType) {
+            // @TODO we'll assume everyting is fine, for now - union and intersection types still need test additions
```
This will be picked up by `infection/infection`, but basically, I did **not** test beyond `ReflectionNamedType` for now.

Ocramius commented on src/DetectChanges/Variance/TypeIsContravariant.php:
```diff
@@ -6,20 +6,26 @@
  * but that is a massive piece of work that should be done by importing an external library
  * instead, if this class no longer suffices.
+ *
+ * @TODO introduce union/intersection type support here
```
TBD

Ocramius commented on src/DetectChanges/Variance/TypeIsCovariant.php:
```diff
@@ -6,21 +6,27 @@
  * but that is a massive piece of work that should be done by importing an external library
  * instead, if this class no longer suffices.
+ *
+ * @TODO introduce union/intersection type support here
```
TBD

Ocramius commented on src/DetectChanges/Variance/TypeIsCovariant.php:
```diff
@@ -31,12 +37,17 @@ public function __invoke(
         }
 
+        if (! $type instanceof ReflectionNamedType || ! $comparedType instanceof ReflectionNamedType) {
+            // @TODO we'll assume everyting is fine, for now - union and intersection types still need test additions
```
This will be picked up by `infection/infection`, but basically, I did **not** test beyond `ReflectionNamedType` for now.

Ocramius committed:
Ensure `IteratorAggregate` is implemented exactly as it is in PHP 8.1 tentative types

Ref: https://wiki.php.net/rfc/internal_method_return_types

Ocramius committed:
Applied automated CS fixes

Ocramius commented:
Test failures are caused by `Psl\Iter\Iterator::current()` not having `: mixed` declared:

```
Deprecated: Return type of Psl\Iter\Iterator::current() should either be compatible with Iterator::current(): mixed, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /github/workspace/vendor/azjezz/psl/src/Psl/Iter/Iterator.php on line 90
```

Problem is: `azjezz/psl:1.9.0` has `php:~8.0.0 || ~8.1.0` support, while the current `azjezz/psl:2.0.0` has `php:~8.1.0` support.

@azjezz do you think we could fix that in `azjezz/psl:1.9.1` perhaps?

azjezz commented:
~1.9.1 is on the way 👍~

done https://github.com/azjezz/psl/releases/tag/1.9.1

Ocramius commented:
Whoa, that was fast, @azjezz and @kukulich! 

I'll work on this some more tonight, trying out both patches ASAP! 💪

Ocramius committed:
Upgraded to `azjezz/psl`, which should fix `php:8.1.0` compatibility

Ref https://github.com/azjezz/psl/releases/tag/1.9.1
Ref https://github.com/azjezz/psl/commit/9fdb771b1d04e57b7ad82fe01a367a37075e47d2

Thanks @azjezz !

Ocramius commented:
Interesting, PHPUnit with lowest dependencies still fails :thinking: 

EDIT:

```
Deprecated: Return type of Symfony\Component\Process\Process::getIterator($flags = 0) should either be compatible with IteratorAggregate::getIterator(): Traversable, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /github/workspace/vendor/symfony/process/Process.php on line 612

Call Stack:
    0.0001     389552   1. {main}() /github/workspace/bin/roave-backward-compatibility-check:0
    0.0006     464800   2. require_once('/github/workspace/bin/roave-backward"
```

I'll add a `"conflict": {"symfony/process": "<5.3.7"}`

Ocramius committed:
Prevent installation with incompatible `symfony/process` versions

Sadly, `symfony/symfony` declared `php:>=7.1` compatibility many moons ago: that is generally
technical lack of vision and responsibility on their end, but that ship has sailed.

This means that a `composer update` can and **will** install a dependency graph that is incompatible
with your system's dependencies.

This change prevents installation of `symfony/process` in versions too archaic to actually function.

As a side-effect, `vimeo/psalm` was also upgraded to its latest and greatest.

Ref: https://github.com/vimeo/psalm/releases/tag/4.12.0

Ocramius commented:
Funny, lowest dependency runs now run forever:

![Selection_824](https://user-images.githubusercontent.com/154256/140625994-e03d69df-411a-4586-9eab-21a1d97e3f26.png)

Ocramius commented on src/DetectChanges/Variance/TypeWithReflectorScope.php:
```diff
@@ -0,0 +1,28 @@
+use Roave\BetterReflection\Reflector\Reflector;
+
+/**
+ * This minimal value object carries an {@see ReflectionType} and the {@see Reflector}
```
This entire file is no longer needed after https://github.com/Roave/BetterReflection/pull/836

Ocramius committed:
Upgraded `roave/better-reflection` to be compatible with https://github.com/Roave/BetterReflection/commit/1f7a13be41d110e132fb8f8ec0fce159c9c4639f

This also handles the upcoming BC break of https://github.com/Roave/BetterReflection/pull/900

Some big changes:

 * `PropertyDocumentedTypeChanged` no longer exists: we use only `PropertyTypeChanged` for now. This
   is a step backwards, but it is needed to prevent issues around docblocks being too advanced for
   this library to parse them, for now.
    * see https://github.com/CuyZ/Valinor/issues/6
    * see https://github.com/phpstan/phpstan-src/tree/447cd102d946a2d9883ac82793195275f54296a9/src/Type
    * see https://github.com/vimeo/psalm/tree/f1d47cc662500589e200a978e3af85488a5236cf/src/Psalm/Type
    * see https://github.com/phpDocumentor/ReflectionDocBlock/blob/622548b623e81ca6d78b721c5e029f4ce664f170/src/DocBlock/Tags/Param.php#L35
    * see https://github.com/phpstan/phpdoc-parser/blob/d639142cf83e91a07b4862217a6f8aa65ee10d08/tests/PHPStan/Parser/TypeParserTest.php#L64
    * /cc @tomasnorre this sadly effectively reverts https://github.com/Roave/BackwardCompatibilityCheck/pull/340 - it
      is too much effort to dive into further advanced types :-(
 * `PropertyTypeChanged` only expects the property to be both covariant and contravariant: if not,
   then a BC break was introduced (this re-uses some internal variance checks, simplifying code
   substantially)
 * usage of `ReflectionFunctionAbstract` is no longer present in the library: always using an union
   of `ReflectionMethod|ReflectionFunction` instead (as in `roave/better-reflection:^5`)
 * `FunctionName` is now `@internal`, since it's only used for internal rendering purposes

Ocramius committed:
Restricting internal types usages to `tests/`, included `psalm/plugin-phpunit`

This change ensures that we don't use `@internal` symbols in `src/` (too dangerous,
for long-term usage), and makes the type checking much stricter, thanks to `psalm/plugin-phpunit`.

Since `vimeo/psalm` has been upgraded, and https://github.com/vimeo/psalm/issues/2772
no longer applies, we can also remove a lot of manually written code that was safe
to clean up.

Ocramius committed:
Implemented covariance/contravariance checks for the new `mixed` and `never` types

Ref: https://wiki.php.net/rfc/mixed_type_v2
Ref: https://wiki.php.net/rfc/noreturn_type

Ocramius committed:
Implemented variance checks for union types

Ref: https://wiki.php.net/rfc/union_types_v2

Ocramius committed:
Implemented variance checks for intersection types

Ref: https://wiki.php.net/rfc/pure-intersection-types

Ocramius committed:
Added tests for nullability + union types variance checks

This change strictly depends on https://github.com/Roave/BetterReflection/pull/902,
and once that's merged, new `infection/infection` mutations should appear, showing
us code that can safely be removed.

Ocramius committed:
Upgraded to latest `roave/better-reflection:5.0.x-dev` to include https://github.com/Roave/BetterReflection/pull/902

Ocramius committed:
Removed all code relying on `ReflectionType#allowsNull()` checks: not necessary with union types

As per previous commit, `roave/better-reflection` has been upgraded so that nullable types are
always represented as union types: this simplifies the logic massively, because, except for
some edge cases, all nullability problems become a set intersection problem, rather than some
ad-hoc conditionals.

This further demonstrates that nullable types as union types is really valuable.

Ref: https://twitter.com/Ocramius/status/1467392834558992390

Ocramius commented:
@WyriHaximus since this is your favorite overall github project, can I have your review too, please? :D

WyriHaximus commented:
@Ocramius Sure, it seems that your CI is failing 😜 . Just kidding, let me go over the thing :)

Ocramius commented:
Haha, yeah, CI is failing because the `6.0.x` branch contains BC breaks (compared to the last `5.0.x` release), so that bit is kinda normal

WyriHaximus commented on src/Changes.php:
```diff
@@ -78,7 +78,7 @@ public function mergeWith(self $other): self
      * @return Traversable<int, Change>
      */
-    public function getIterator(): iterable
+    public function getIterator(): Traversable
```
It is more of a learning moment for myself: but why did you make this change?

Ocramius replied:
Oh, good point, this could stay iterable! 🤔

Ocramius replied:
Checked: this is necessary for PHP 8.1 compatibility, where `IteratorAggregate#getIterator(): Traversable` is declared.

The change is therefore valid, and shouldn't pose a downstream BC break, since `Traversable` is covariant with `iterable` here

WyriHaximus reviewed (approved):
Not perse my overall favourite project but I'm using it for some cool stuff https://github.com/WyriHaximus/php-test-utilities/pull/260/files#diff-c8f1537d7020a410a1f19233721bb24de6aa93df4ef9b40795453c1647ed5f99R31 and some other stuff is blocked by it https://github.com/WyriHaximus/php-test-utilities/pull/332

WyriHaximus commented:
@Ocramius Ow yeah, I didn't even look at that. Just wanted to joke about it :P. Jokes aside, this is one reason why I want to use this package more: Make it comment on PR's about what BC would be introduced where.

Ocramius commented:
:ship:
